src/Microsoft.ML.AutoML/Experiment/SuggestedPipeline.cs

```diff
@@ -118,8 +119,13 @@ public IEstimator<ITransformer> ToEstimator()
                 }
             }
 
+            if(validationSet != null)
+            {
+                validationSet = pipeline.Fit(trainset).Transform(trainset);
```

Comment:
I was tracking down how you were creating a validation set, to ensure we're not overfitting the validation dataset. We'd have to work out the impact of re-using the existing validation dataset (as it serves the purpose of a test set in this code). The actual test set is kept separately besides for codegen.

I read the current code as: If there is a validation dataset, replace it with a featurized version of the training set. If there is no validation dataset, keep it empty. (bug?)

Creating another validation set by featurizing the trainset is not bad from a re-use of splits reason, but you may not get what you're looking for. I think you're using it within your trainer for early stopping. 

If you used the actual validation dataset for early stopping, this would skew the user's metrics by over-representing the quality of the model.

I'd recommend creating a new TrainTest split to create an independent dataset split for the early stopping. You may want this as a feature of your trainer itself.

Reply:
Nope, the validation set in AutoML code is unchanged, this is just a validation set that is passed to the image classification trainer for its own internal calculations. 

---
In reply to: [339943032](https://github.com/dotnet/machinelearning/pull/4395#discussion_r339943032) [](ancestors = 339943032)

Reply:
P.S: early stopping is disabled by default.

Reply:
Since I saw this has moved from `validationSet = pipeline.Fit(trainset).Transform(trainset)` to `validationSet = pipeline.Fit(validationSet).Transform(validationSet)`. 

This now uses what the code is calling a validationSet, though it is currently the test dataset as used here. We shouldn't be reusing the dataset for this purpose. This hits on my earlier comment of:
> If you used the actual validation dataset for early stopping, this would skew the user's metrics by over-representing the quality of the model.

Even if not used currently for early stopping, other uses of this dataset will have similar effects.

As mentioned before, we should split off the training set:
> I'd recommend creating a new TrainTest split to create an independent dataset split for the early stopping. You may want this as a feature of your trainer itself.

Reply:
Define reuse? Can you please walk me through how this is being used anywhere in the trainer, I have put it for a reason that I can explain.